src/workerd/api/worker-rpc.c++

```diff
@@ -2,19 +2,387 @@
 // Licensed under the Apache 2.0 license found in the LICENSE file or at:
 //     https://opensource.org/licenses/Apache-2.0
 
-#pragma once
-
 #include <workerd/api/worker-rpc.h>
+#include <workerd/io/features.h>
+#include <workerd/api/global-scope.h>
 
 namespace workerd::api {
 
+namespace {
+kj::Array<kj::byte> serializeV8(jsg::Lock& js, jsg::JsValue value) {
+  auto outcome = js.tryCatch([&]() -> kj::OneOf<kj::Array<kj::byte>, kj::String> {
+    jsg::Serializer serializer(js, jsg::Serializer::Options {
+      .version = 15,
+      .omitHeader = false,
+    });
+    serializer.write(js, value);
+    auto released = serializer.release();
+    return kj::mv(released.data);
+  }, [&](jsg::Value exception) {
+    return kj::str(jsg::check(exception.getHandle(js)->ToString(js.v8Context())));
+  });
+
+  KJ_SWITCH_ONEOF(outcome) {
+    KJ_CASE_ONEOF(e, kj::String) {
+      JSG_FAIL_REQUIRE(Error, e);
+    }
+    KJ_CASE_ONEOF(arr, kj::Array<kj::byte>) {
+      return kj::mv(arr);
+    }
+  }
+  KJ_UNREACHABLE;
+}
+
+jsg::JsValue deserializeV8(jsg::Lock& js, kj::Array<const kj::byte> ser) {
+  jsg::Deserializer deserializer(js, kj::mv(ser), kj::none, kj::none,
+      jsg::Deserializer::Options {
+    .version = 15,
+    .readHeader = true,
+  });
+
+  return deserializer.readValue(js);
+}
+} // namespace
+
+kj::Promise<WorkerRpc::ResponseAndResult> WorkerRpc::sendWorkerRpc(
+    jsg::Lock& js,
+    kj::StringPtr name,
+    const v8::FunctionCallbackInfo<v8::Value>& args) {
+
+  auto& ioContext = IoContext::current();
+  auto worker = getClient(ioContext, kj::none, "getJsRpcTarget"_kjc);
+  auto event = kj::heap<api::GetJsRpcTargetCustomEventImpl>(WORKER_RPC_EVENT_TYPE);
+
+  rpc::JsRpcTarget::Client client = event->getCap();
+  auto builder = client.callRequest();
+  builder.setMethodName(name);
+
+  kj::Vector<jsg::JsValue> argv(args.Length());
+  for (int n = 0; n < args.Length(); n++) {
+    argv.add(jsg::JsValue(args[n]));
+  }
+
+  // If we have arguments, serialize them.
+  if (argv.size() > 0) {
+    auto ser = serializeV8(js, js.arr(argv.asPtr()));
+    builder.initSerializedArgs().setV8Serialized(kj::mv(ser));
+  }
```

Comment:
Keep in mind that v8 serialization does not cover a lot of types still. Anything that is backed by a JavaScript class, for instance, or any of our API types (e.g. `URL`), will not be properly supported here. We'll want to document that there are definite limits to the types of arguments that can be passed.

Reply:
Good point, I'll add some comments in the code + keep that in mind when we write docs. I do test for this in our internal repo fwiw (confirms we get an error message).